src/transformers/modeling_utils.py

```diff
@@ -1474,8 +1814,12 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
 
             if low_cpu_mem_usage:
                 # save the keys
-                loaded_state_dict_keys = [k for k in state_dict.keys()]
-                del state_dict  # free CPU memory - will reload again later
+                if is_sharded:
+                    loaded_state_dict_keys = sharded_metadata["all_checkpoint_keys"]
+                else:
+                    state_dict = load_state_dict(resolved_archive_file)
+                    loaded_state_dict_keys = [k for k in state_dict.keys()]
+                    del state_dict  # free CPU memory - will reload again later
```

Comment:
Same here, adapts the logic.

`sharded_metadata["all_checkpoint_keys"]` is built in the function above to contain all the keys that are present in the index of the checkpoint (and thus in the union of all shards).